drivers/lte_link_control/lte_lc.c

```diff
@@ -153,23 +162,51 @@ static int w_lte_lc_init(void)
 
 static int w_lte_lc_connect(void)
 {
-	if (at_cmd_write(network_mode, NULL, 0, NULL) != 0) {
-		return -EIO;
-	}
+	int err;
+	char *current_network_mode = (char *)nw_mode_preferred;
+	bool use_fallback = false;
 
+	at_cmd_set_notification_handler(at_handler);
 	k_sem_init(&link, 0, 1);
 
-	at_cmd_set_notification_handler(at_handler);
+retry:
+	if (IS_ENABLED(CONFIG_LTE_NETWORK_USE_FALLBACK) &&
+	    (use_fallback != false)) {
+		current_network_mode = (char *)nw_mode_fallback;
+
+		LOG_INF("Using fallback network mode");
+
+		if (at_cmd_write(offline, NULL, 0, NULL) != 0) {
+			return -EIO;
+		}
+	}
+
+	LOG_DBG("Network mode: %s", log_strdup(current_network_mode));
+
+	if (at_cmd_write(current_network_mode, NULL, 0, NULL) != 0) {
+		return -EIO;
+	}
 
 	if (at_cmd_write(normal, NULL, 0, NULL) != 0) {
 		return -EIO;
 	}
 
-	k_sem_take(&link, K_FOREVER);
+	err = k_sem_take(&link, K_SECONDS(CONFIG_LTE_NETWORK_TIMEOUT));
+	if (err == -EAGAIN) {
+		LOG_INF("Network connection attempt timed out.");
+
+		if (IS_ENABLED(CONFIG_LTE_NETWORK_USE_FALLBACK) &&
+		    (use_fallback == false)) {
+			use_fallback = true;
+			goto retry;
```

Comment:
Not a strong request if you have another opinion, but...

While I'm all for using `goto` statements to reduce return points from a function and simplify cleanups, I'm not sure if we should overuse it for other purposes. Here, for instance, it's used as a replacement for a regular loop, and using a simple `do { ... } while(retry);` loop here would make the code a little easier to follow IMO.

Reply:
I was thinking about something like this:
```
do {
	retry = false;
	... try to connect and wait ...
	if (err = -EAGAIN) {
		if (IS_ENABLED(CONFIG_LTE_NETWORK_USE_FALLBACK) &&
		    current_network_mode == nw_mode_preferred) {
			... set fallback mode and do whatever cleanup is needed ...
			retry = true;
		} else {
			err = -ETIMEDOUT
		}
	}
} while (retry);
```

Reply:
Used `goto` to save indentation. But your suggestion is easier to read and maintain, so switched to `do { ... } while ()`

Reply:
Yep, char limit was my concern as well with loop design (4 indentations), still I find such a flow easier to follow, even if some extra line breaks would be needed - fortunately, they were not.